python/pyarrow/tests/test_dataset.py

```diff
@@ -374,7 +376,127 @@ def test_partition_scheme_discovery(mockfs):
         ("group", pa.int32()),
         ("key", pa.string()),
     ])
-    assert inspected_schema.remove_metadata().equals(expected_schema)
+    assert inspected_schema.equals(expected_schema, check_metadata=False)
 
     hive_discovery = ds.HivePartitionScheme.discover()
     assert isinstance(hive_discovery, ds.PartitionSchemeDiscovery)
+
+
+def test_partitioning_function():
+    schema = pa.schema([("year", pa.int16()), ("month", pa.int8())])
+    names = ["year", "month"]
+
+    # default SchemaPartitionScheme
+
+    part = ds.partitioning(schema)
+    assert isinstance(part, ds.SchemaPartitionScheme)
+    part = ds.partitioning(names)
+    assert isinstance(part, ds.PartitionSchemeDiscovery)
+    # needs schema or names
+    with pytest.raises(ValueError):
+        ds.partitioning()
+
+    # Hive partitioning
+
+    part = ds.partitioning(schema, flavor="hive")
+    assert isinstance(part, ds.HivePartitionScheme)
+    part = ds.partitioning(flavor="hive")
+    assert isinstance(part, ds.PartitionSchemeDiscovery)
+    # cannot pass list of names
+    with pytest.raises(ValueError):
+        ds.partitioning(names, flavor="hive")
+
+    # unsupported flavor
+    with pytest.raises(ValueError):
+        ds.partitioning(schema, flavor="unsupported")
+
+
+@pytest.mark.parquet
+def test_open_dataset_single_file(tempdir):
+    import pyarrow.parquet as pq
+    table = pa.table({'a': range(9), 'b': [0.] * 4 + [1.] * 5})
+    pq.write_table(table, tempdir / "test.parquet")
+
+    dataset = ds.dataset(str(tempdir / "test.parquet"))
+    assert dataset.schema.equals(table.schema, check_metadata=False)
+
+    result = dataset.new_scan().finish().to_table()
+    result = result.replace_schema_metadata()
+    assert result.equals(table)
+
+
+@pytest.mark.parquet
+def test_open_dataset_pathlib(tempdir):
+    import pyarrow.parquet as pq
+    import pathlib
+
+    table = pa.table({'a': range(9), 'b': [0.] * 4 + [1.] * 5})
+    pq.write_table(table, tempdir / "test.parquet")
+
+    path = tempdir / "test.parquet"
+    assert isinstance(path, pathlib.Path)
+    dataset = ds.dataset(path)
+    result = dataset.new_scan().finish().to_table()
+    assert result.replace_schema_metadata().equals(table)
+
+
+@pytest.mark.parquet
+def test_open_dataset_directory(tempdir):
```

Comment:
There's a lot of repeated code in these tests. Please parameterize them or write helpers

Reply:
Can you specify a bit more? What aspect do you find repetitive? 

The creation of the data? (but this is different for each test) Or the way that the resulting dataset is checked? That could indeed be written with a helper function, but it might also point to a too verbose API (eg @kszucs was adding a Dataset.to_table method so `dataset.new_scan().finish().to_table()` could be shortened to `dataset.to_table()`)

Reply:
Here's a sketch of some helpers and a rewrite of `test_open_dataset_directory` which should clarify what I mean. I won't block on this; the test suite isn't super long yet

For data creation:
```py
def write_parquet_table(columns, path):
    assert isinstance(path, pathlib.Path)
    path.parent.mkdir(exist_ok=True)

    table = pa.table(columns)

    import pyarrow.parquet as pq
    pq.write_table(table, path)
    return table
```

For dataset extraction:
```py
def dataset_table(path, **kwargs):
    assert isinstance(path, pathlib.Path)
    dataset = ds.dataset(path, **kwargs)
    table = dataset.to_table()
    # verify equivalently creatable from string path
    assert ds.dataset(str(path)).to_table().equals(table)
    return dataset, table


@pytest.mark.parquet
def test_open_dataset(tempdir):
    columns = {'a': range(9), 'b': [0.] * 4 + [1.] * 5}
    expected = write_parquet_table(columns, tempdir / 'test.parquet')
    dataset, actual = dataset_table(tempdir / 'test.parquet')
    assert actual.replace_schema_metadata().equals(expected)
```